New bitheader type "bounds", pre/post columns and few quality of life things

nopeslide opened the pull request:
I added the bitheader type "bounds" which is similar to the "smart" one.
It shows the first and last index of a field. 

I also added the possibility for columns before and after the bytefield i.e. for indexes and labels,
These can be filled with tablex primitives which are now passed through. 

closes #1

nopeslide committed:
added bitheader type "bounds" and few quality of life things

nopeslide commented on example.typ:
```diff
@@ -1,4 +1,4 @@
-#import "@local/bytefield:0.0.3": *
+#import "bytefield.typ" : *
```
could not compile without changing this

jomaway replied:
Yeah don't mind. I added a symlink inside the local typst package folder to the repo folder to check if it works as a package.

nopeslide committed:
added numbers to label bitheader

jomaway commented:
Thanks for the PR. I will have a look at it soon. Please be a bit patient.

nopeslide commented on bytefield.typ:
```diff
@@ -96,15 +112,37 @@
           #if (h_text != "" and bitheader.at("marker", default: auto) != none){ place(bottom, line(end:(0pt, 5pt))) }
           #rotate(bitheader.at("angle", default: -60deg), origin: left, h_text)
+          ]
+          #if (type(numbers) == bool and numbers and h_text != "") {
+              v(-0.5em)
+              align(center, text(bitheader_font_size)[#i])
```
just noticed that this is undocumented in the examples.
setting `numbers: true` adds numbers to a bit if it has a specified label

nopeslide committed:
fixup! added bitheader type "bounds" and few quality of life things

nopeslide committed:
fixup! added numbers to label bitheader

nopeslide committed:
refactored examples

nopeslide committed:
fixup! refactored examples

jomaway commented:
@nopeslide found some time today to look into your PR. The "bound" thing looks good so far. 

I also like your other ideas. Like the `pre` and `post` column. Tested it with `rowspanx` which works out of the box. This is great, but if you want them only in certain lines you need to insert blank fields which i don't like that much. Causes some "unwanted" output if you forget them or if they are too long. 
![Bildschirmfoto vom 2024-02-08 23-23-12](https://github.com/jomaway/typst-bytefield/assets/22931557/5ab48140-f711-493d-a9c9-7c431e44a110)

I think we can keep it like this for the moment and improve later on. So if you don't have some quick idea on this issues i will merge your changes on the weekend.

jomaway commented on example.typ:
```diff
@@ -1,82 +1,134 @@
+#import "bytefield.typ" as bf
+#import "@preview/codelst:2.0.0": sourcecode
+
+#let example(columns:2,source) = box(grid(
```
Please change to `#let example(columns: 1,source)`  for the default.

nopeslide replied:
Isn't side-by-side the better choice, because it is more compact and therefore makes the example more understandable?

nopeslide committed:
fixed bitbox cellx positions

nopeslide commented:
I fixed the x/y positions of the cellx elements in tablex. that lets you omit empty pre/post fields.

nopeslide committed:
introduced annotation field for easier placement with note/group helper

nopeslide commented:
Introduced a new field "annotation" similar to "bitbox" to make the placement of labels in multiple pre/post columns easier.
Also added a `note` and `group` helper to show off :D

jomaway commented on bytefield.typ:
```diff
@@ -189,15 +203,56 @@
+  args: args,
+  body: body
+)
+
```
Nice, i had the same idea. Because then we can swap out tablex to the new buildin table implementation if it is ready without changing the bytefield api. 

Could you shortly describe what pos and col refers to?

nopeslide replied:
uh sorry, force pushed a better example just this minute.
`pos` is the position left/right of the bytefield in the pre/post columns.
`col` specifies the exact column if you have multiple pre/post columns (see the new example)

nopeslide replied:
How about renaming these?
`pos` to `side`
`col` to simply `y` or `column`  ?

nopeslide replied:
`col` could also be a `level`
if the columns is calculated as distance from the bytefield. So it becomes more of hierarchical notion
`side:left` calculates `pre.len() - level - 1`
And
`side:right` calculates `pre.len() + bits + level`

jomaway commented:
I have something in the pipeline to convert the bitbox in a bitcell which contains information about start and end bit to make header bounds and smart header easier, as it doesn't have to be calculated seperately.  Just not finished yet, as i testing some designs with multi rows bytefields.

nopeslide commented:
> I have something in the pipeline to convert the bitbox in a bitcell which contains information about start and end bit to make header bounds and smart header easier, as it doesn't have to be calculated seperately. Just not finished yet, as i testing some designs with multi rows bytefields.

Nice.
This would make border handling easier,
especially if you also annotate their position (like start/end/middle in the bitbox)
I would love borders like in [bytefield-svg](https://bytefield-svg.deepsymmetry.org/bytefield-svg/1.8.0/funcs.html).

nopeslide committed:
renamed arguments, added examples, automated pre/post

nopeslide committed:
fixed breakup if annotation is overflowing

nopeslide commented:
@jomaway sorry for my chaotic PR
I will stop adding stuff and wait for your feedback

jomaway commented:
> @jomaway sorry for my chaotic PR I will stop adding stuff and wait for your feedback

Don't worry. Thanks for all the ideas and improvements. I will take the weekend to merge your PR and make some changes from my side as well.

nopeslide commented:
I think the next thing that should be done is a refactor.
my next ideas:
- make a generic two-pass system (or even three-pass)
  - simplifies main bytefield function to only run passes and pipe final result into tablex
  - generic in terms of it only calls functions with a fixed signature, has no own logic 
    - have a dict of type to callback mappings, i.e. `callbacks = ("bitbox" : bitbox_callback, "annotations": annotation_callback, ..)`
    - simply execute the dict `callbacks.at(field.type)(field)`
  - first pass is a fold over the supplied fields, generating a dict
  - second pass maps the supplied field with the generated dict into internal cells
  - third pass maps the internal cells to tablex cells
    - would keep the internal representation slim and on spot. sth like bfcell to cellx 
- make bitheader a custom field like annotation
  - labels and numbers can be split into two bitheader fields
  - allows special bitheader in between, at the bottom etc
